### ISSUE TITLE
Order of loading of global middlewares

### ISSUE THREAD

**Issue by adnan-kamili** [pleerock/routing-controllers#202]
I have two global middlewares (both useBefore), how is the order of execution decided if we load the middlewares from directory.

Secondly, order of middlewares if passed as an array doesn't seem to have any effect. 

```javascript
const app = createKoaServer({
    defaultErrorHandler: false,
    routePrefix: "/v1",
    authorizationChecker: async (action: Action, claims: string[]) => {
        const verifyJwt = jwt(jwtConfig).unless({ path: ['/v1/auth/token', /^\/v1\/accounts/] });
        await verifyJwt(action.context, () => { });
        const user = action.context.state.user;
        if (user.role.includes('admin')) {
            return true;
        }
        for (const claim of claims) {
            if (user.scope.includes(claim)) {
                return true;
            }
        }

        throw new ForbiddenError("you do not have required permissions");
    },
    controllers: [__dirname + "/api/controllers/*.js"],
    middlewares: [CustomErrorHandler, JwtMiddleware]
}).listen(3000);
```

**Comment by NoNameProvided** [comment]
> how is the order of execution decided if we load the middlewares from directory.

The `@Middleware` decorator supports the `priority` option. ([src/decorator/Middleware.ts#L7](https://github.com/pleerock/routing-controllers/blob/master/src/decorator/Middleware.ts#L7)) The lower is the higher.

>  order of middlewares if passed as an array doesn't seem to have any effect.

That should work, when you pass them as an array, they should be executed in the same order. In your case `CustomErrorHandler` handler will be executed first and `JwtMiddleware` as second. (Dont you want to use your `CustomErrorHandler` as last? An error dropped in the `JwtMiddleware` wont be cached by it if you call it first.)

**Comment by adnan-kamili** [comment]
CustomErrorHandler wraps await next() inside a try catch hence it catches every error, so should be invoked first to wrap the whole pipeline in the try catch block. Changing the order of middlewares in the array didn't have any effect, JwtMiddleware was always invoked first. I think priority needs to be added to make it work even for array.

**Comment by 19majkel94** [comment]
> I think priority needs to be added to make it work even for array.

You're right, I've checked that and without `priority` the order is unknown. Only after adding `priority` the order is correct as we've set.

The key is in that line:
https://github.com/pleerock/routing-controllers/blob/master/src/RoutingControllers.ts#L98

We sort the array by priority but the sorting algorithm in V8 is not stable so it's get randomized when `priority` is undefined (for middlewares loaded explicitly, not by dir glob) or because of `.reverse();` call the order is reversed. Also the middlewares from dir are concated to the explicit array:
https://github.com/19majkel94/routing-controllers/blob/404-return-type-fix/src/index.ts#L159-L161

So @NoNameProvided I think we could filter the middlewares that doesn't have `priority` (manually loaded) and then concat to them the rest sorted by priority. If this makes sense, I will try to make an PR 😉

**Comment by NoNameProvided** [comment]
Hmm, I think we should just swap them for a default value instead of handling them differently. 

But there is more than that to consider. 

### Lower is higher vs Higher is Higher

It's strange to have a lower is higher approach on priority. I think this should be swapped. But this is a breaking change. What do you think about this? (cc @pleerock because this is a breaking change, but we are below 1.0 so in theory we are allowed to do it.) 

### Sorting algorithm

We should write a custom sort function to prevent calling reverse after sorting. Also if we would have a "Higher is Higher" approach we would be able to default to `1` when priority is `undefined`. 

If we make those change a basic sorting function could be something like:

```ts
 .sort((ma, mb) => {
  ma.priority = ma.priority ? ma.priority : 1;
  mb.priority = mb.priority ? mb.priority : 1;

  return mb.priority - ma.priority;
})
```

Thoughts @19majkel94 & @pleerock

**Comment by 19majkel94** [comment]
> It's strange to have a lower is higher approach on priority. I think this should be swapped. But this is a breaking change. What do you think about this? (cc @pleerock because this is a breaking change, but we are below 1.0 so in theory we are allowed to do it.)

We have this in code:
```ts
.sort((middleware1, middleware2) => middleware1.priority - middleware2.priority)
.reverse()
```
So `3, 1, 5` is sorted as `1, 3, 5` but then reversed, so higher priority value -> first in middleware chain. And I've just checked that and confirm that higher is higher, which is correct 😃 

The problem is with reverse sorting - V8 sorting algo isn't stable so we get reverse order, not the one we defined in array:
```ts
const objects = [
    {
        name: "und one",
    },
    {
        name: "und two",
    },
    {
        name: "und three",
    },
    {
        name: "four",
        value: 4,
    },
    {
        name: "two",
        value: 2,
    },
    {
        name: "one",
        value: 1,
    },
    {
        name: "three",
        value: 3,
    },
]

const sorted1 = objects.sort((a, b) => a.value - b.value).reverse()
console.log(sorted1)
```
```
​​​​​[
  { name: 'four', value: 4 },​​​​​
​​​​​  { name: 'three', value: 3 },​​​​​
​​​​​  { name: 'two', value: 2 },​​​​​
​​​​​  { name: 'one', value: 1 },​​​​​
​​​​​  { name: 'und three' },​​​​​
​​​​​  { name: 'und two' },​​​​​
​​​​​  { name: 'und one' },
]​​​​​
```
So we need to copy the middlewares passed explicitly, sort only the glob-string-loaded ones and concat it to the first one. So we should receive then:
```
[
  { name: 'und one' },
  { name: 'und two' },
  { name: 'und three' },​​​​​
  { name: 'four', value: 4 },​​​​​
​​​​​  { name: 'three', value: 3 },​​​​​
​​​​​  { name: 'two', value: 2 },​​​​​
​​​​​  { name: 'one', value: 1 },​​​​​
]​​​​​
```
If you agree with me, I think I could to the PR 😉

**Comment by NoNameProvided** [comment]
> So 3, 1, 5 is sorted as 1, 3, 5 but then reversed, so higher priority value -> first in middleware chain.

😱  then I just lied in my first comment, sorry for that @adnan-kamili!

> The problem is with reverse sorting 

But we dont have to call reverse if we sort it correctly the first time. All you have to do is call

```
const sorted = objects.sort((a, b) => b.value - a.value)
console.log(sorted)
```

**Comment by 19majkel94** [comment]
You're right 😔 I've forget that the sorting mutates the array, not like the filter returning new array but only a reference to the mutated array:
```js
const sorted1 = objects.sort((a, b) => a.value - b.value).reverse()
console.log(sorted1)
const sorted2 = objects.sort((a, b) => b.value - a.value)
console.log(sorted2)
```
And it printed the same. But now I've checked and you're right. Undefined is not a problem in that case:
```js
const sorted2 = [...objects].sort((a, b) => b.value - a.value)
console.log(sorted2)
```
```
​​​​​[ { name: 'und one' },​​​​​
​​​​​  { name: 'und two' },​​​​​
​​​​​  { name: 'und three' },​​​​​
​​​​​  { name: 'four', value: 4 },​​​​​
​​​​​  { name: 'three', value: 3 },​​​​​
​​​​​  { name: 'two', value: 2 },​​​​​
​​​​​  { name: 'one', value: 1 } ]​​​​​
```
So it should be really small fix. However `sort()` isn't guaranteed to be stable  - looks like it is in V8 but I don't know how node-chakra behave :wink:

**Comment by NoNameProvided** [comment]
>  However sort() isn't guaranteed to be stable

Thats true, and we can overcome it by simply defaulting to the lowest priority if priority is `undefined`.

```ts
 .sort((ma, mb) => {
  ma.priority = ma.priority ? ma.priority : 1;
  mb.priority = mb.priority ? mb.priority : 1;

  return mb.priority - ma.priority;
})
```

**Comment by 19majkel94** [comment]
> A sorting algorithm is said to be stable if two objects with equal keys appear in the same order in sorted output as they appear in the input unsorted array

So priority 1 is the same as undefined. But no, priority 1 means that middleware will be called on the end of the chain.

**Comment by NoNameProvided** [comment]
> Oh, it's even worse because priority 1 means that middleware will be called on the end of the chain.

It's not worse, its by design, all the middlewares are priority 1 (or maybe even zero?) which is the lowest by design and will execute in the default order. If you want to prioritize a middleware simply raise it's priority.

We cannot use `undefined` anyway because: `undefined - 1` becomes `NaN`. What is not good.

But when they are added as array,  we should maintain the order between middlewares with default priority as they were added.

**Comment by 19majkel94** [comment]
So we go to the main problem - what is the priority of middleware without priority. Right now we can mix explicit class and glob strings in routing-controllers options.

In the code middlewares that are loaded from directories are placed after the explicit ones. So without priority option they should be called last?

But what if someone define priority -1? It should be called after the explicit, and 2 should be before? I don't know why 1 is neutral.

I have no idea what should be the order of mixed middlewares. For me neutral should be 0 and we should write it in docs.

**Comment by adnan-kamili** [comment]
Usually the number of  middlewares is less than the number of models and controllers, and the idea of allowing them to load then from a path, imho doesn't seem to be a good decision. If only allowed through array, the problem of priority is automatically resolved, as it could be the order of middlewares in the array, and it sounds more intuitive.

**Comment by NoNameProvided** [comment]
> But what if someone define priority -1? It should be called after the explicit, and 2 should be before? I don't know why 1 is neutral.
>
> I have no idea what should be the order of mixed middlewares. For me neutral should be 0 and we should write it in docs.

No we should not allow numbers below zero. We can use 0 as default, that is totally fine for me.

> So we go to the main problem - what is the priority of middleware without priority.

and 

> and we should write it in docs.

We just defined it, all middleware without priority should default to the neutral priority. (whatever it is 1 or 0)

> and the idea of allowing them to load then from a path, imho doesn't seem to be a good decision.

I totally agree with that, I would like to see them only loaded explicitly as an array. As loading order may vary between machines / os-es when loaded from a path.

While loading them from a path as a barrel export looks almost the same as loading them from a path. But **they have always an explicit order.** 

> Property keys are traversed in the following order:
>
> - First, the keys that are integer indices (what these are is explained later), in ascending numeric order.
> - Then, all other string keys, in the order in which they were added to the object.
> - Lastly, all symbol keys, in the order in which they were added to the object.
- [ref](http://2ality.com/2015/10/property-traversal-order-es6.html#traversing-the-own-keys-of-an-object)

So the code could be something like this:

```ts
import * as MIDDLEWARES from './middlewars';

useExpressServer(this.express, {
      authorizationChecker,
      defaultErrorHandler: false,
      middlewares: Object.keys(MIDDLEWARES).map(key => (<any>MIDDLEWARES)[key]),
      routePrefix: '',
    });

```

We use this right now. However this was downvoted by @19majkel94  and @pleerock before.

**Comment by 19majkel94** [comment]
```ts
import * as MIDDLEWARES from './middlewars';
(...)
Object.keys(MIDDLEWARES).map(key => (<any>MIDDLEWARES)[key]),
```
For me it's the same as import from directories using glob string. You have no control of middlewares, you can't disable single middleware for debugging/development purposes.

BTW, `Object.keys(MIDDLEWARES).map(key => (<any>MIDDLEWARES)[key])` isn't just `Object.values(MIDDLEWARES)`?

**Comment by NoNameProvided** [comment]
> For me it's the same as import from directories using glob string. You have no control of middlewares, you can't disable single middleware for debugging/development purposes.

You can: 

```ts
// middlewares/index.ts
// order is explicit
export * from 'x.middleware.ts';
// this is disabled
// export * from 'y.middleware.ts';
export * from 'z.middleware.ts';
```

> BTW, Object.keys(MIDDLEWARES).map(key => (<any>MIDDLEWARES)[key]) isn't just Object.values(MIDDLEWARES)?

Indeed it is.

**Comment by 19majkel94** [comment]
I think that index file is a win-win. You have explicit priority and ability to temporarily disable selected one + clean code on initialization (`useExpressServer` location) - no 20+ lines with imports and then joining together in one array. I have to use this pattern in typeorm entities too 😆 

One drawback I can see is that can break your unit test so you may get false negative due to error with importing sth to test.

**Comment by 19majkel94** [comment]
Also we should consider accepting also object where keys implements MiddlewareInterface. So no need for `Object.keys(MIDDLEWARES).map(key => (<any>MIDDLEWARES)[key])`, just:
```ts
useExpressServer(this.express, {
    middlewares: MIDDLEWARES,
}
```

**Comment by NoNameProvided** [comment]
> I think that index file is a win-win.

I am happy to see that you have changed your mind, I wanted to introduce this back in https://github.com/pleerock/routing-controllers/issues/131#issuecomment-301255798

So lets move forward with this! 😎

**Comment by 19majkel94** [comment]
> I wanted to introduce this back in #131 (comment) what was the reason back in then to not support it? 

Here are the reasons:
> Yeah, but we are going back to the problem - the order of middleware is undetermined, object property iteration order is not guaranteed. So we again need a decorator to configure it.

> I would consider Object.keys(Middlewares).map(key => Middlewares[key]) a hack to use if u really dont care about the order (Is there even any scenario when you really dont care?), but I would consider it a bad practice what should be avoided. (Just like loading middlewares from path.)

But since you proved that `string keys, in the order in which they were added to the object` so we can control the order/priority in index.ts file, it's time to implement it.

I would also like to implement this feature to get rid of the need of `@Middleware()` decorator and setting there the `type` parameter. It would be better to set it in index file or import and place them in array:
```ts
createExpressServer({
    middlewares: {
        before: beforeMiddlewares, // or middlewares.before
        after: afterMiddlewares,  // or middlewares.after
        error: [
            CustomErrorMiddleware,
        ]
    }
});
```

**Comment by 19majkel94** [comment]
> so we have to map the items in array and give them descending priority values to make it works

Looks like it's not so easy:
```ts
    filterMiddlewareMetadatasForClasses(classes: Function[]): MiddlewareMetadataArgs[] {
        return this.middlewares.filter(ctrl => {
            return classes.filter(cls => ctrl.target === cls).length > 0;
        });
    }
```
We have to modify this part of the code to return `MiddlewareMetadataArgs` matching to the order of classes(middlewares). Any ideas @NoNameProvided? The only workaround is to mark the classes much earlier as the "direct from array" and then change the priority prop of `MiddlewareMetadata`.

**Comment by NoNameProvided** [comment]
Ups forgot to reply to this one. Did you figure it out? Or should I take a look as well? I have no ideas without digging in. 😄

**Comment by 19majkel94** [comment]
The key is in the code pasted up here - we have to modify this part of the code to return MiddlewareMetadataArgs matching to the order of classes(middlewares). I have no idea how to make it nicely, not by dirty hacks.

However I would like to introduce loading-by-object feature with separate `before` and `after` option like described before. It would be easier to make a big refactor of this part with fix and new feature in one PR. I have your confirmation but I'm afraid pleerock is not a fan of this solution because it may require dropping support for glob string or introduce more complications in the codebase.

**Comment by NoNameProvided** [comment]
> I have your confirmation but I'm afraid pleerock is not a fan of this solution

It's a more elegant solution, so I am sure he will see the value in it also.

> We should create a separate issue for this proposal to track the discussion as this issue will be closed soon.

Yes please do that and reference back this issue.
